core/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -240,61 +261,83 @@ public void beforeListVariableChanged(ListVariableDescriptor<Solution_> variable
     public void afterListVariableChanged(ListVariableDescriptor<Solution_> variableDescriptor, Object entity, int fromIndex,
             int toIndex) {
         var notifiables = notifiableRegistry.get(variableDescriptor);
+        var notification = Notification.<Solution_> listVariableChanged(entity, fromIndex, toIndex);
         if (!notifiables.isEmpty()) {
-            ListVariableNotification<Solution_> notification = Notification.listVariableChanged(entity, fromIndex, toIndex);
             for (var notifiable : notifiables) {
                 notifiable.notifyAfter(notification);
             }
             notificationQueuesAreEmpty = false;
         }
-        listVariableEventList.add(new ListVariableEvent(entity, fromIndex, toIndex));
+        if (!cascadingUpdateShadowVarDescriptorList.isEmpty()) { // Only necessary if there is a cascade.
+            listVariableChangedNotificationList.add(notification);
+        }
     }
 
     public void triggerVariableListenersInNotificationQueues() {
         for (var notifiable : notifiableRegistry.getAll()) {
             notifiable.triggerAllNotifications();
         }
-        if (listVariableDescriptor != null && !cascadingUpdateShadowVarDescriptorList.isEmpty()) {
+        // If there is no cascade, skip the whole thing.
+        // If there are no events and no newly unassigned variables, skip the whole thing as well.
+        if (!cascadingUpdateShadowVarDescriptorList.isEmpty() &&
+                !(listVariableChangedNotificationList.isEmpty() && unassignedValueWithEmptyInverseEntitySet.isEmpty())) {
             triggerCascadingUpdateShadowVariableUpdate();
         }
         notificationQueuesAreEmpty = true;
-        listVariableEventList.clear();
     }
 
-    /**
-     * Triggers all cascading update shadow variable user-logic.
-     */
     private void triggerCascadingUpdateShadowVariableUpdate() {
-        if (listVariableEventList.isEmpty() || cascadingUpdateShadowVarDescriptorList.isEmpty()) {
-            return;
-        }
         for (var cascadingUpdateShadowVariableDescriptor : cascadingUpdateShadowVarDescriptorList) {
-            for (var event : listVariableEventList) {
-                var values = listVariableDescriptor.getValue(event.entity());
-                // Evaluate all elements inside the range
-                evaluateFromIndex(values, event.fromIndex(), event.toIndex(), true, cascadingUpdateShadowVariableDescriptor);
-                // Evaluate later elements, but stops when there is no change
-                evaluateFromIndex(values, event.toIndex(), values.size(), false, cascadingUpdateShadowVariableDescriptor);
```

Comment:
Both of these separate iterations have been merged together to just one iteration.
If you can iterate once, iterate once; useful for CPU caches.